numpy/core/src/multiarray/item_selection.c

```diff
@@ -1749,6 +1750,281 @@ PyArray_LexSort(PyObject *sort_keys, int axis)
     return NULL;
 }
 
+/*
+* Helper function for PyArray_KeySort
+* Sorts 1D slice based on argsort indices
+* takes in raw bytes of data, indices array
+* Number of elements, and stride of data
+* element size and temporary storage for an element.
+*/
+void reorder_by_index(char *data, char *current, npy_intp *index, npy_intp N, npy_intp stride, npy_intp elsize) {
+    // Reorder the indices and relevant values cyclically
+    for(npy_intp i = 0; i < N; i++){
+        /*Backup current element, end of loop place it in correct location.
+        *Stride here handles both aligned and non-aligned case
+        *When stride is same as elsize we are operating on aligned data.
+        */
+        memmove(current, data + (i*stride), elsize);
+        npy_intp current_idx = i;
+        // Break when index and index buffer are the  same.
+        while(1) {
+           npy_intp next_idx = index[current_idx];
+           index[current_idx] = current_idx;
+           if (next_idx == i) {
+              break;
+           }
+           memmove(data +(current_idx*stride), data + (next_idx*stride), elsize);
+           current_idx = next_idx;
+        }
```

Comment:
I am wondering if rewriting this as a `do { } while ()` loop may be more iomatic? Not quite sure though.